File: 2018/04-sleeping-guards/go/star2.go
package main

import (
    "io/ioutil"
    "fmt"
    "strings"
    "sort"
)

func main() {
    bytes, _ := ioutil.ReadFile("input")
    lines := strings.Split(string(bytes), "\n")
    lines = lines[:len(lines)-1]
    sort.Strings(lines)

    var year, month, day, hour int // only used for correct parsing
    var guard, minute, sleepStart int
    minutesSlept := make(map[[2]int]int)
    for _, line := range lines {
        fmt.Sscanf(line, "[%d-%d-%d %d:%d]", &year, &month, &day, &hour, &minute)
        action := line[19:]

        switch {
            case action[0:1] == "G":
                fmt.Sscanf(action, "Guard #%d begins shift", &guard)
            case action == "falls asleep":
                sleepStart = minute
            case action == "wakes up":
                for i := sleepStart; i < minute; i++ {
                    guardMinute := [2]int{guard, i}
                    minutesSlept[guardMinute] += 1
                }
        }
    }

    var sleepTime int
    var sleepiest [2]int
    for guardMinute, minutes := range minutesSlept {
        if minutes > sleepTime {
            sleepiest = guardMinute
            sleepTime = minutes
        }
    }

    fmt.Println(sleepiest[0] * sleepiest[1])
}

File: 2018/04-sleeping-guards/rust/star2.rs
use std::collections::HashMap;

const INPUT: &str = include_str!("../input");

fn parse(line: &str) -> (&str, u32) {
    let splitted: Vec<&str> = line.split(']').collect();
    let action = &splitted[1][1..];
    let minute: u32 = splitted[0][15..].parse().unwrap();
    return (&action, minute);
}

fn measure_sleep_times(sorted_lines: Vec<&str>) -> HashMap<(u32, u32), u32> {
    let mut guard_minute_sleep_time: HashMap<(u32, u32), u32> = HashMap::new();

    let mut current_guard = 0;
    let mut sleep_start = 0;
    for line in sorted_lines {
        let (action, minute) = parse(line);
        if action.starts_with("Guard #") {
            current_guard = action.split(' ').nth(1).unwrap()[1..].parse().unwrap();
        } else if action == "falls asleep" {
            sleep_start = minute;
        } else if action == "wakes up" {
            for min in sleep_start..minute {
                *guard_minute_sleep_time
                    .entry((current_guard, min))
                    .or_default() += 1;
            }
        }
    }

    return guard_minute_sleep_time;
}

fn main() {
    let mut lines: Vec<&str> = INPUT.lines().collect();
    lines.sort();

    let guard_minute_sleep_time = measure_sleep_times(lines);

    let mut sleepiest_guard_minute: (&(u32, u32), &u32) = (&(0, 0), &0);
    for guard_minute in guard_minute_sleep_time.iter() {
        if guard_minute.1 > sleepiest_guard_minute.1 {
            sleepiest_guard_minute = guard_minute;
        }
    }
    let sleepiest_guard_minute = sleepiest_guard_minute.0;

    println!("{:?}", sleepiest_guard_minute.0 * sleepiest_guard_minute.1);
}
